### PR TITLE
Fix: Hotkey equip on cpp + fixes

### LINKED ISSUE #175: Equiping Ammo Bug - Action Bars
This one is very tricky:

When you equip arrows/bolts, give that some requirements are fulfilled, 100 of your ammo will be thrown to the ground, beneath you.

1. Create a Paladin with level 150+ or a GM/Admin for the following tests
2. Have 100 power bolts in your arrow slot, and in your backpack, in this exact order:
 - 99 infernal bolts, 74 infernal bolts, 15 diamond arrows and 100 diamond arrows
3. Put the infernal bolts and diamond arrows on your action bar, set to equip.
4. Switch to infernal bolts through action bar
5. A stack of 100 should have been placed at your feet
6. Now with the 100 power bolts still in your arrow slot, switch to diamond arrows, again, through action bar
7. 100 arrows should be laying beneath you, 1 by 1 (because their decay time will be different, for some reason lol)

image
image
image
image
image
image
image
image
image
image

The ammount of ammo does not matter, considering that you have more than 100 and the first stack of it is less than 100.

### LINKED ISSUE #1039: Equiping Ammo Bug - Action Bars
This one is very tricky:

When you equip arrows/bolts, give that some requirements are fulfilled, 100 of your ammo will be thrown to the ground, beneath you.

1. Create a Paladin with level 150+ or a GM/Admin for the following tests
2. Have 100 power bolts in your arrow slot, and in your backpack, in this exact order:
 - 99 infernal bolts, 74 infernal bolts, 15 diamond arrows and 100 diamond arrows
3. Put the infernal bolts and diamond arrows on your action bar, set to equip.
4. Switch to infernal bolts through action bar
5. A stack of 100 should have been placed at your feet
6. Now with the 100 power bolts still in your arrow slot, switch to diamond arrows, again, through action bar
7. 100 arrows should be laying beneath you, 1 by 1 (because their decay time will be different, for some reason lol)

image
image
image
image
image
image
image
image
image
image

The ammount of ammo does not matter, considering that you have more than 100 and the first stack of it is less than 100.

### PR DESCRIPTION
**Pull request by marcosvf132**
Resolves #1039
Resolves some parts of #836 (Are the other part really bug?)

- [x] Hotkey equip/dequip changed to CPP.
- [x] No delay on equip/dequip item (As global tibia do).
- [x] Restack ammo if pressed hotkey item is the same as slot.
- [x] Diamond arrow/spectral bolt can be stacked only if there are equal. (Preventing reset decay time)
- [x] Ammo are sent to the ammo quickloot backpack, if not possible , to the main backpack. (As global tibia do)

Thank to @dracoalv for guiding me how it should be working.

### CODE REVIEW

**Comment by dracoalv** [comment]
A hotkey equip ficou perfeita para arrow e bolts no geral com excecão das Diamonds Arrows e Spectral Bolts cujo não completam o inventário em caso de faltas, nada grave.

File: src/protocolgame.cpp [hunk 1]
```diff
@@ -495,6 +495,7 @@ void ProtocolGame::parsePacket(NetworkMessage& msg)
 		case 0x71: addGameTaskTimed(DISPATCHER_TASK_EXPIRATION, &Game::playerTurn, player->getID(), DIRECTION_SOUTH); break;
 		case 0x72: addGameTaskTimed(DISPATCHER_TASK_EXPIRATION, &Game::playerTurn, player->getID(), DIRECTION_WEST); break;
 		case 0x73: parseTeleport(msg); break;
+		case 0x77: onPressHotkeyEquip(msg); break;
```
Comment (FakeShinoda) [inline src/protocolgame.cpp]: Can you make this a **parsePressHotkeyEquip** and create a function just for parse?

File: src/protocolgame.cpp [hunk 2]
```diff
@@ -586,6 +586,219 @@ void ProtocolGame::parsePacket(NetworkMessage& msg)
 		}
 	*/
 }
+void ProtocolGame::onPressHotkeyEquip(NetworkMessage& msg)
```
Comment (FakeShinoda) [inline src/protocolgame.cpp]: With the parse function.. You can create a game.cpp function to handle this code, use protocolgame.cpp just for communications

File: src/protocolgame.cpp [hunk 3]
```diff
@@ -586,6 +586,219 @@ void ProtocolGame::parsePacket(NetworkMessage& msg)
...
+
+	if (itemType.id == 0) {
+		return;
+	}
+
+	bool removed = false;
+	ReturnValue ret = RETURNVALUE_NOERROR;
+	item = g_game.findItemOfType(player, itemType.id);
+
+	if (item == nullptr) {
```
Comment (FakeShinoda) [inline src/protocolgame.cpp]: you can also try !item

**Review by FakeShinoda — CHANGES_REQUESTED** [review]

File: src/protocolgame.cpp [hunk 4]
```diff
@@ -586,217 +586,14 @@ void ProtocolGame::parsePacket(NetworkMessage& msg)
...
-		  ret = g_game.internalMoveItem(item->getParent(), player, slotP, item, item->getItemCount(), nullptr);
-		}
-	  }
-	}
-
-	if (ret != RETURNVALUE_NOERROR) {
-	  player->sendCancelMessage(ret);
-	}
-
+	g_game.onPressHotkeyEquip(player, spriteid);
```
Comment (FakeShinoda) [inline src/protocolgame.cpp]: Better use **addGameTask(&Game::onPressHotkeyEquip, player, spriteid);**

**Review by FakeShinoda — APPROVED** [review]

**Review by dudantas — APPROVED** [review]

**Review by majestyotbr — APPROVED** [review]

**Pull request merged**